### DISCUSSION TITLE
A new and more complete user-interface.

### DISCUSSION THREAD

**Discussion by marco-sacchi** [marco-sacchi/RLoader#31] [Work in progress]
Given some requests from [this topic](https://www.vogons.org/viewtopic.php?f=24&t=81788) I decided to make the RLoader user interface more complete.

This is a development step that will take more time than the previous work, and it is not possible to get an idea of what it will look like from the source code alone, so you will find screenshots here.

If you have any questions or would like to give me feedback they will be welcome.

Text-based user interface:
![image](https://user-images.githubusercontent.com/81907055/163682255-e13ffbe9-c8da-4bfa-bcd0-ba16e4072dc2.png)
_A view (equivalent of a .NET form) built by instantiating controls. Accelerators that make accessing the UI faster are visible._

![image](https://user-images.githubusercontent.com/81907055/163682271-317cb4e6-c1d3-4bf2-ae66-d234ac54f607.png)
_A message box with automatic sizing/centering and text wrapping._

![image](https://user-images.githubusercontent.com/81907055/163682301-59d46717-18aa-45bd-ac11-6ae007dd6fc3.png)
_Standard open/save/select view._

![image](https://user-images.githubusercontent.com/81907055/163682891-404a1c4a-1f6d-4680-b6da-6c669dc74a71.png)
_List items highlight._

Graphics user interface (into the same state as text-based above):
![image](https://user-images.githubusercontent.com/81907055/163683589-385ae950-46de-4674-9784-47b2d8dd4bbd.png)
![image](https://user-images.githubusercontent.com/81907055/163682433-27c8c7bc-fb8b-41db-8a2f-92edd60e4d34.png)
![image](https://user-images.githubusercontent.com/81907055/163682505-128c6abe-a01e-4635-9d68-afdf60466ddf.png)
![image](https://user-images.githubusercontent.com/81907055/163682850-3106142d-801a-4470-9ea2-4fc19da34722.png)

All the controls you see are functional and follow the standard of the first graphical interfaces for DOS.

You can navigate with the following keys:
- <kbd>Tab</kbd> and <kbd>Shift</kbd> + <kbd>Tab</kbd>: Move to previous / next control with
- <kbd>Alt</kbd> + <kbd>[Character]</kbd>: activate or move directly to the control with corresponding "accelerator" key
- Arrow keys, <kbd>PgUp</kbd>/<kbd>PgDn</kbd>, <kbd>Home</kbd>/<kbd>End</kbd>: move the cursor by character in the input fields or move between the items contained in a control
- <kbd>Ctrl</kbd> + Arrows keys: move the cursor by word in the input fields
- <kbd>Space</kbd>: activate a control or an item of a control
- <kbd>Alt</kbd> + <kbd>F3</kbd>: close a view, when it is not modal
- [Characters...]: from within a list it is possible to directly jump to an item by typing the first characters

The graphical interface is 1-bit color like the old one, because the rest of the palette will be used by thumbnails.

There are still some improvements to be implemented, but this extension gives me the possibility to arrange the interface more easily and to develop other features more quickly, such as being able to modify the program list directly from within RLoader.

**Comment by cyberluke** [comment]
Very nice!

**Comment by vicokoby** [comment]
What tools do you use to make the thumbnails?

**Reply by marco-sacchi** [reply]
@vicokoby first of all sorry for the late reply, and thanks for using rloader!

You can use any command line tool that supports image manipulation, for example using [ImageMagick](https://imagemagick.org/), which is a free and open source tool, you can write:

```
magick.exe input.png -resize 320 -depth 4 -colors 14 bmp3:output.bmp
```

where:
- `-resize 320` scale down the image to 320 pixels
- `-depth 4` specify 4-bits per pixels (16 colors)
- `-colors 14` remap using only 14 colors, a limit required by rloader because 2 colors must be reserved for the UI
- `bmp3` means windows uncompressed bitmap

You can avoid bilinear filtering caused by downscaling by adding the `-filter` flag with the values ​​`Point` or `NearestNeighbor`.

I did a test just now using version 7.1.1 portable of ImageMagick. The output image is loaded correctly by rloader.

**EDIT**

I added a reference section in the [readme](https://github.com/marco-sacchi/RLoader?tab=readme-ov-file#thumbnails-generation).

**Comment by vicokoby** [comment]
@marco-sacchi
Using those options with ImageMagick, I'm getting 24-bit BMPs, and I also can't use the NearestNeighbor filter.
Are you sure you're using ImageMagick version 1.7.7?

**Reply by marco-sacchi** [reply]
@vicokoby yes, I have tested it before replying to you, downloading the latest portable version of ImageMagick for Windows, and opening the thumbnail into rloader.

**Reply by vicokoby** [reply]
Maybe the problem is the source images I'm using, I got these from https://gamesdb.launchbox-app.com/games/details/13749-the-lion-king
![lionking2](https://github.com/user-attachments/assets/cbca9f3f-45f4-4ada-aa8b-18b252e04ef1)
![lionking](https://github.com/user-attachments/assets/7dad1543-6498-4723-8c18-dffa57470648)
Edit: I'm adding the BMP outputs:
[BMPs.ZIP](https://github.com/user-attachments/files/19839368/BMPs.ZIP)

**Reply by marco-sacchi** [reply]
My input images were in 256 colors (8-bit per pixel), to properly convert 24-bit per pixel images I verified that this command works correctly:

```
.\magick.exe .\input.png -resize 320 -dither Ordered -depth 4 -colors 14 -filter Point -type palette bmp3:output.bmp
```

You can omit the `-dither` flag if you want. If you use it, make sure that your version of ImageMagick supports the dithering method, and pay attention to the order of the parameters.

Let me know if you still have problems.

**Reply by vicokoby** [reply]
ImageMagick can't seem to return 4-bit BMPs from 32-bit images. I've had better luck with 16-bit images, and with these new options, the thumbnails are more accurate.

Thanks for your help.
